src/setup_remote.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+import subprocess, os, sys, re, time
+
+this_dir = os.path.dirname(__file__)
+SLURM_GCP_DOCKER_DIR = os.path.realpath(os.path.join(this_dir, ".."))
+
+def download_getzlab_ssh_key():
+    pubkey = os.path.join(SLURM_GCP_DOCKER_DIR, "getzlabkey.pub")
+    seckey = os.path.join(SLURM_GCP_DOCKER_DIR, "getzlabkey")
+    if not os.path.exists(pubkey):
+        subprocess.check_call(["gsutil", "cp", "gs://getzlab-secrets/github-service-account/github.pub", pubkey])
+    if not os.path.exists(seckey):
+        subprocess.check_call(["gsutil", "cp", "gs://getzlab-secrets/github-service-account/github", seckey])
+
+def get_current_project():
+    ## Assuming on a GCE instance
+    project = subprocess.check_output(
+        "curl -s -H 'Metadata-Flavor: Google' http://metadata.google.internal/computeMetadata/v1/project/project-id",
+        shell=True
+    )
+    project = project.decode()
+    return project
+
+def get_current_instance_name():
+    name = subprocess.check_output(
+        "curl -s -H 'Metadata-Flavor: Google' http://metadata.google.internal/computeMetadata/v1/instance/name",
+        shell=True
+    )
+    name = name.decode()
+    return name
+
+def get_current_zone():
+    ## Assuming on a GCE instance
+    zone = subprocess.check_output(
+        "curl -s -H 'Metadata-Flavor: Google' http://metadata.google.internal/computeMetadata/v1/instance/zone",
+        shell=True
+    )
+    zone = zone.decode()
+    zone = re.sub(".*/", "", zone)
+    return zone
+
+def create_instance(instance_name, project=None, zone=None, machine_type="n1-standard-4", boot_disk_size=200):
+    if project is None:
+        project = get_current_project()
+    if zone is None:
+        zone = get_current_zone()
+    ## I used "enable-oslogin=TRUE", which ensures consistent UID within a project.
+    subprocess.check_call(
+        f"gcloud compute instances create {instance_name} \
+            --quiet \
+            --project {project} \
+            --zone {zone} \
+            --machine-type {machine_type} \
+            --image ubuntu-minimal-2004-focal-v20210511 \
+            --image-project ubuntu-os-cloud \
+            --boot-disk-size {boot_disk_size}GB \
+            --boot-disk-type pd-standard \
+            --metadata=enable-oslogin=TRUE \
+            --scopes cloud-platform,compute-rw"
+            #--metadata-from-file startup-script={startup_script},shutdown-script={shutdown_script} \
+            , shell=True
+    )
+
+    ## Waiting for ssh working
+    n = 0
+    while True:
+        time.sleep(6)
+        try:
+            subprocess.check_call(f"gcloud compute ssh --project {project} --quiet {instance_name} --zone {zone} --command 'true'", shell=True, stderr=subprocess.DEVNULL)
+            break
+        except Exception as err:
+            if n >= 5:
+                raise err
+            pass
+        finally:
+            n += 1
```

Comment:
Should we add exponential backoff to the timeout, and give the user a friendly error in addition to the exception trace?